spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteringTableFeatureSuite.scala

```diff
@@ -49,24 +61,86 @@ class ClusteringTableFeatureSuite extends SparkFunSuite with DeltaSQLCommandTest
         sql("ALTER TABLE tbl CLUSTER BY (a)")
       }
       checkError(
         e1,
-        "DELTA_ALTER_TABLE_CLUSTER_BY_NOT_ALLOWED",
+        "DELTA_ALTER_TABLE_CLUSTER_BY_ON_PARTITIONED_TABLE_NOT_ALLOWED",
         parameters = Map.empty)
 
       val e2 = intercept[DeltaAnalysisException] {
         sql("ALTER TABLE tbl CLUSTER BY NONE")
       }
       checkError(
         e2,
-        "DELTA_ALTER_TABLE_CLUSTER_BY_NOT_ALLOWED",
+        "DELTA_ALTER_TABLE_CLUSTER_BY_ON_PARTITIONED_TABLE_NOT_ALLOWED",
         parameters = Map.empty)
     }
   }
+
+   test("alter table cluster by unpartitioned tables is supported.") {
+    val table = "tbl"
+    withTable(table) {
+      sql(s"CREATE TABLE $table (a INT, b STRING) USING DELTA")
+      val (_, startingSnapshot) = DeltaLog.forTableWithSnapshot(spark, TableIdentifier(table))
+      assert(!ClusteredTableUtils.isSupported(startingSnapshot.protocol))
+      val clusterByLogs = Log4jUsageLogger.track {
+        sql(s"ALTER TABLE $table CLUSTER BY (a)")
+      }.filter { e =>
+        e.metric == MetricDefinitions.EVENT_TAHOE.name &&
+          e.tags.get("opType").contains("delta.ddl.alter.clusterBy")
+      }
+      assert(clusterByLogs.nonEmpty)
+      val clusterByLogJson = JsonUtils.fromJson[Map[String, Any]](clusterByLogs.head.blob)
+      assert(!clusterByLogJson("isClusterByNoneSkipped").asInstanceOf[Boolean])
+      val (_, finalSnapshot) = DeltaLog.forTableWithSnapshot(spark, TableIdentifier(table))
+      assert(ClusteredTableUtils.isSupported(finalSnapshot.protocol))
+      val dependentFeatures = TableFeature.getDependentFeatures(ClusteringTableFeature)
+      dependentFeatures.foreach { feature =>
+        assert(finalSnapshot.protocol.isFeatureSupported(feature))
+      }
+
+      withSQLConf(SQLConf.MAX_RECORDS_PER_FILE.key -> "2") {
+        val df = (1 to 4).map(i => (i, i.toString)).toDF("a", "b")
+        withSQLConf(SQLConf.MAX_RECORDS_PER_FILE.key -> "1") {
+          df.write.format("delta").mode("append").saveAsTable(table)
+        }
+
+        // Optimize should cluster the data into two 2 files since MAX_RECORDS_PER_FILE is 2.
+        runOptimize(table) { metrics =>
+          assert(metrics.numFilesRemoved == 4)
+          assert(metrics.numFilesAdded == 2)
+        }
+
+        val deltaLog = DeltaLog.forTable(spark, TableIdentifier(table))
+        val files1 = deltaLog.update().allFiles.collect().toSet
+        assert(files1.size == 2)
+        assert(files1.forall(_.clusteringProvider.contains(ClusteredTableUtils.clusteringProvider)))
+
+        val tableLocation = deltaLog.dataPath.toString
+        // Check if min-max intervals of 'a' are sorted
+        val minMaxIntervals = files1.map { file =>
+          val df = spark.read.format("parquet").load(s"$tableLocation/${file.path}")
+          val minMax = df.agg(min("a").as("min_a"), max("a").as("max_a")).collect().head
+          (minMax.getAs[Int]("min_a"), minMax.getAs[Int]("max_a"))
+        }
```

Comment:
Ah, you can just get the min and max from stats of AddFiles.

Reply:
ah, updated